### PR TITLE
Get public key

### PR DESCRIPTION
**Pull request by cool-user**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -4,7 +4,9 @@
       "env",
       {
         "targets": {
-          "browsers": ["last 2 versions"],
+          "browsers": [
```
Comment (ppershing) [inline .babelrc]: I am not sure if we can modify this if we want to merge back to original ledgerjs repository

File: packages/example-node/src/index.js [hunk 2]
```diff
@@ -1,12 +1,18 @@
 import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
-import AppBtc from "@ledgerhq/hw-app-btc";
+import AppAda from "@ledgerhq/hw-app-ada";
```
Comment (ppershing) [inline packages/example-node/src/index.js]: Again, this probably should not be merged back to ledgerjs

File: packages/hw-app-ada/src/Ada.js [hunk 3]
```diff
@@ -22,7 +22,7 @@ import { TransportStatusError } from "@ledgerhq/hw-transport";
 
 const CLA = 0x80;
 
-const INS_GET_PUBLIC_KEY = 0x01;
+const INS_GET_EXT_PUBLIC_KEY = 0x10;
 const INS_SET_TX = 0x02;
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js]: rest of these is obsolete now ...

File: packages/hw-app-ada/src/Ada.js [hunk 4]
```diff
@@ -57,8 +57,7 @@ export default class Ada {
     this.transport = transport;
     this.methods = [
       "isConnected",
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js]: let's rename this to `getVersion`. (Please cross-check with BTC what naming do they use)

File: packages/hw-app-ada/src/Ada.js [hunk 5]
```diff
@@ -87,78 +86,47 @@ export default class Ada {
...
-   * @description Get a public key from the specified BIP 32 index.
-   * The BIP 32 index is from the path at `44'/1815'/0'/[index]`.
-   *
-   * @param {number} index The index to retrieve.
-   * @return {Promise<{ publicKey:string }>} The public key for the given index.
-   *
-   * @throws 5201 - Non-hardened index passed in, Index < 0x80000000
-   * @throws 5202 - Invalid header
-   * @throws 5003 - Index not a number
+   * @throws 5003 - Some of the indexes is not a number
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:83]: I gues this doc comment contains obsolete info by now

File: packages/hw-app-ada/src/Ada.js [hunk 6]
```diff
@@ -87,78 +86,47 @@ export default class Ada {
...
    */
-  async getWalletPublicKeyWithIndex(
-    index: number
-  ): Promise<{ publicKey: string }> {
-    if (isNaN(index)) {
+  async getExtendedPublicKey(
+    indexes: Array<number>
+  ): Promise<{ publicKey: string, chainCode: string }> {
+    if (indexes.some(index => isNaN(index))) {
       throw new TransportStatusError(INDEX_NAN);
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:97]: Is this compatible with how BTC/other currencies check the validity of input? For that matter, do they expect Array<number> ?

File: packages/hw-app-ada/src/Ada.js [hunk 7]
```diff
@@ -87,78 +86,47 @@ export default class Ada {
...
     }
 
-    const data = Buffer.alloc(4);
-    data.writeUInt32BE(index, 0);
+    const data = Buffer.alloc(1 + 4 * indexes.length);
+    data.writeUInt8(indexes.length, 0);
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:104]: I guess we should check that `indexes.length` fits in uint8

File: packages/hw-app-ada/src/Ada.js [hunk 8]
```diff
@@ -87,78 +86,47 @@ export default class Ada {
...
       CLA,
-      INS_GET_PUBLIC_KEY,
-      0x02,
+      INS_GET_EXT_PUBLIC_KEY,
+      0x00,
       0x00,
       data
     );
 
     const [publicKeyLength] = response;
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js]: Oh, does this get the first value out of uint8 array?

File: packages/hw-app-ada/src/Ada.js (hunk 4 above)
Comment (cool-user) [reply packages/hw-app-ada/src/Ada.js]: From my fast look at the BTC app, looks like they dont have app version call.

File: packages/hw-app-ada/src/Ada.js (hunk 5 above)
Comment (cool-user) [reply packages/hw-app-ada/src/Ada.js:83]: Nope, it was updated.

File: packages/hw-app-ada/src/Ada.js [hunk 9]
```diff
@@ -16,31 +16,18 @@
...
-const P2_MULTI_TX = 0x02;
-
-const MAX_APDU_SIZE = 64;
-const OFFSET_CDATA = 5;
-const MAX_ADDR_PRINT_LENGTH = 12;
-const INDEX_MAX = 0xffffffff;
+const INS_GET_VERSION = 0x00;
+const INS_GET_EXT_PUBLIC_KEY = 0x10;
 
+const INVALID_PATH = 0x5001;
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:29]: Are these from ledger or they are pure JS exception values. I guess we should distinguish them somehow (for now at least by comment)

File: packages/hw-app-ada/src/Ada.js [hunk 10]
```diff
@@ -56,227 +43,83 @@ export default class Ada {
...
-   * @throws 5201 - Non-hardened index passed in, Index < 0x80000000
-   * @throws 5202 - Invalid header
-   * @throws 5003 - Index not a number
+   * @throws 5001 - The path provided does not have the first 3 indexes hardened
+   * @throws 5002 - The path provided is less than 3 indexes
+   * @throws 5003 - Some of the indexes is not a number
    *
    * @example
-   * const { publicKey } = await ada.getWalletPublicKeyWithIndex(0xC001CODE);
+   * const { publicKey, chainCode } = await ada.getExtendedPublicKey([ 0x80000000 + 44, 0x80000000 + 1815, 0x80000000 + 1 ]);
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js]: we have now constant HARDENED

File: packages/hw-app-ada/src/Ada.js [hunk 11]
```diff
@@ -56,227 +44,83 @@ export default class Ada {
...
-    index: number
-  ): Promise<{ publicKey: string }> {
-    if (isNaN(index)) {
+  async getExtendedPublicKey(
+    indexes: Array<number>
+  ): Promise<{ publicKey: string, chainCode: string }> {
+    if (indexes.length < 3 || indexes.length > 10) {
+      throw new TransportStatusError(INVALID_PATH_LENGTH);
+    }
+    if (indexes.some(index => isNaN(index))) {
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:96]: Could be done later: The check probably should be strenthtened to check for exact 32bit unsigned ints ....

File: packages/hw-app-ada/src/Ada.js [hunk 12]
```diff
@@ -56,227 +44,83 @@ export default class Ada {
...
-      INS_GET_PUBLIC_KEY,
-      0x02,
+      INS_GET_EXT_PUBLIC_KEY,
+      0x00,
       0x00,
       data
     );
 
-    const [publicKeyLength] = response;
+    const publicKeyLength = response[0];
```
Comment (ppershing) [inline packages/hw-app-ada/src/Ada.js:118]: Just a note: this does not correspond to current design doc. Obviously, we would need to fix it in the future

**Review by ppershing — APPROVED** [review]

**Pull request merged**